sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java

```diff
@@ -663,7 +663,16 @@ long flush(
                   retrieveErrorDetails(contexts));
               failedContext.failureCount += 1;
 
-              invalidateWriteStream();
+              Throwable error = Preconditions.checkStateNotNull(failedContext.getError());
+              Status.Code statusCode = Status.fromThrowable(error).getCode();
+              boolean quotaError = statusCode.equals(Status.Code.RESOURCE_EXHAUSTED);
+
+              if (!quotaError) {
+                // This forces us to close and reopen all gRPC connections to Storage API on error,
+                // which empirically
+                // fixes random stuckness issues.
+                invalidateWriteStream();
```

Comment:
invalidateWriteStream is currently just unpinning.  Should it close() as well?
In the recent issue we appeared to have leaks of ManagedChannels in AppendClient, could that have been because they were not closed there?

Reply:
close() is called on cache eviction, and invalidateWriteStream also evicts from the cache. Cache owns the object, so we only want to close on cache eviction.